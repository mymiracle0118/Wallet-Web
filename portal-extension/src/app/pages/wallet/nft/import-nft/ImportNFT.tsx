import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { CustomTypography, Input, ModalComponent, Form, Button } from 'components'
import SinglePageTitleLayout from 'layouts/single-page-layout/SinglePageLayout'
import mascot from 'assets/images/mascot.png'
import { useWallet, Asset } from '@portal/shared/hooks/useWallet'
import { isNFTOwner, fetchNFT } from '@portal/shared/services/nft'
import { useNavigate } from 'lib/woozie'
import { default as networks } from '@portal/shared/data/networks.json'

import { useForm } from 'react-hook-form'
import { yupResolver } from '@hookform/resolvers/yup'
import * as yup from 'yup'
import { SpinnerIcon } from '@src/app/components/Icons'

const schema = yup.object().shape({
  contractAddress: yup.string().required('Please enter a valid address'),
  // .matches(/^0x[a-fA-F0-9]{40}$/, 'Invalid address'),
  tokenId: yup.string().min(1).max(10).required('Please enter your token Id'),
})

const ImportNFT = () => {
  const { t } = useTranslation()
  const { navigate } = useNavigate()
  const [isLoading, setIsLoading] = useState<boolean>(false)
  const [isOwner, setIsOwner] = useState<boolean>(false)
  const [modalState, setModalState] = useState<boolean>(false)
  // eslint-disable-next-line
  const [preview, setPreview] = useState<Asset | null>(null)
  const { address, addNFT } = useWallet()

  const methods = useForm({
    mode: 'onChange',
    resolver: yupResolver(schema),
  })
  const {
    handleSubmit,
    register,
    getValues,
    formState: { errors, isValid },
  } = methods

  const getPreview = async () => {
    if (isValid) {
      setIsLoading(true)

      const networksToCheck = networks.filter((n) => [1, 3, 5, 42].includes(n.chainId as number))
      for (const network of networksToCheck) {
        try {
          const res = await fetchNFT(
            network.networkId,
            getValues('contractAddress') as string,
            getValues('tokenId') as string
          )

          if (res) {
            setPreview(res)
            break
          }
        } catch (error) {
          console.log(error)
        }
      }

      setIsLoading(false)
    }
  }

  useEffect(() => {
    if (preview) {
      isNFTOwner(address as string, preview.network, preview.contractAddress as string, preview.id)
        .then(setIsOwner)
        .catch((err) => console.log(err))
    }
  }, [address, preview])

  return (
    <SinglePageTitleLayout title="Import NFT">
      <Form
        methods={methods}
        onSubmit={handleSubmit(() => {
          if (preview) {
            addNFT(preview)
            setModalState(true)
          }
        })}
      >
        <div className="flex flex-col p-4 relative h-[500px]">
          <CustomTypography variant="body" className="mb-4">
            You can find the below information in the Details section on the OpenSea page of your NFT
          </CustomTypography>
          <Input
            fullWidth
            mainColor
            placeholder="Contract address"
            className="mt-4"
            {...register('contractAddress', {
              onBlur: getPreview,
            })}
            error={errors.contractAddress?.message}
          />
          <Input
            className="mt-4"
            fullWidth
            mainColor
            placeholder="ID"
            {...register('tokenId', {
              onBlur: getPreview,
            })}
            error={errors.tokenId?.message}
          />
          {preview && !isOwner && (
            <CustomTypography variant="subtitle" className="mt-4" color="text-feedback-negative">
              You do not own this NFT
            </CustomTypography>
          )}

          {preview && (
            <div className="flex gap-4 mt-4">
              <img src={preview?.metadata?.image} alt="nft" className="rounded-md w-40" />
              <div className="flex flex-col gap-2">
                <CustomTypography variant="h3" className="mb-4">
                  {preview?.metadata?.name}
                </CustomTypography>
                <CustomTypography className="text-subtitle1" type="secondary">
                  {preview?.description?.slice(0, 140)}...
                </CustomTypography>
              </div>
            </div>
          )}
          <div className="flex gap-2 absolute bottom-0 w-auto right-4 left-4">
            <Button variant="bordered" color="outlined">
              {t('Actions.cancel')}
            </Button>
            <Button
              type="submit"
              color={`${!isOwner || !isValid ? 'disabled' : 'primary'}`}
              isDisabled={!isOwner || !isValid}
            >
              Import
            </Button>
          </div>
        </div>
      </Form>
      <ModalComponent
        modalState={modalState}
        closeModal={() => setModalState(false)}
        image={mascot}
        title={`You have imported ${preview?.metadata?.name as string}`}
        modalIconImage={preview?.metadata?.image}
      >
        <Button onClick={() => navigate('/home')} isLoading={isLoading} spinner={<SpinnerIcon />} color="primary">
          {!isLoading && t('Actions.ok')}
        </Button>
      </ModalComponent>
    </SinglePageTitleLayout>
  )
}

export default ImportNFT
