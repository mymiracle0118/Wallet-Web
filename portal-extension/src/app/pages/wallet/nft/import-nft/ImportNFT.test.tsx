import React from 'react'
import { render } from '@testing-library/react'
import ImportNFT from 'pages/wallet/nft/import-nft/ImportNFT'
import { FormProvider, useForm } from 'react-hook-form'

const WrapperForm = ({ children }) => {
  const methods = useForm()
  return <FormProvider {...methods}>{children}</FormProvider>
}
describe('<ImportNFT />', () => {
  test('renders correctly', () => {
    const tree = render(
      <WrapperForm>
        <ImportNFT />
      </WrapperForm>
    )
    expect(tree).toMatchSnapshot()
  })
})
