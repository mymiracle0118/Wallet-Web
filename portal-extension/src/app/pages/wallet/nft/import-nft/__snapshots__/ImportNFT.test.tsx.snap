// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImportNFT /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black"
      >
        <div
          class="flex flex-row relative justify-between items-center h-16"
        >
          <button
            class="h-[24px] w-[24px] p-1 ml-1 hover:bg-custom-grey10 dark:hover:bg-custom-white10 rounded-full flex items-center justify-center absolute"
            type="button"
          >
            <arrow-left.svg
              classname="text-xl stroke-custom-black dark:stroke-custom-white"
            />
          </button>
          <div
            class="flex-1 text-center"
          >
            <div
              class="text-custom-black dark:text-custom-white text-lg font-extrabold"
            >
              Import NFT
            </div>
          </div>
        </div>
        <div
          class="w-full h-full overflow-auto"
        >
          <form>
             
            <div
              class="flex flex-col p-4 relative h-[500px]"
            >
              <div
                class="text-custom-black dark:text-custom-white text-sm font-regular"
                style="margin-bottom: 4px;"
              >
                You can find the below information in the Details section on the OpenSea page of your NFT
              </div>
              <div
                class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10 mt-4"
              >
                <input
                  autocomplete="off"
                  class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                  name="contractAddress"
                  placeholder="Contract address"
                  value=""
                />
              </div>
              <div
                class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10 mt-4"
              >
                <input
                  autocomplete="off"
                  class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                  name="tokenId"
                  placeholder="ID"
                  value=""
                />
              </div>
              <div
                class="flex gap-2 absolute bottom-0 w-auto right-4 left-4"
              >
                <button
                  class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full h-[44px] border-[1px] border-solid text-primary dark:text-custom-white border-primary dark:border-custom-white40 dark:hover:border-custom-white hover:border-primary-alt hover:text-primary-alt"
                  type="button"
                >
                  Actions.cancel
                </button>
                <button
                  class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full disabled:text-custom-grey40 dark:disabled:text-custom-grey40 h-[44px] text-custom-white dark:text-custom-white bg-custom-grey10"
                  disabled=""
                  type="submit"
                >
                  Import
                </button>
              </div>
            </div>
          </form>
          <div
            style="position: absolute; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
          />
        </div>
      </div>
    </div>
    <div
      id="headlessui-portal-root"
    >
      <div
        data-headlessui-portal=""
      >
        <div />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black"
    >
      <div
        class="flex flex-row relative justify-between items-center h-16"
      >
        <button
          class="h-[24px] w-[24px] p-1 ml-1 hover:bg-custom-grey10 dark:hover:bg-custom-white10 rounded-full flex items-center justify-center absolute"
          type="button"
        >
          <arrow-left.svg
            classname="text-xl stroke-custom-black dark:stroke-custom-white"
          />
        </button>
        <div
          class="flex-1 text-center"
        >
          <div
            class="text-custom-black dark:text-custom-white text-lg font-extrabold"
          >
            Import NFT
          </div>
        </div>
      </div>
      <div
        class="w-full h-full overflow-auto"
      >
        <form>
           
          <div
            class="flex flex-col p-4 relative h-[500px]"
          >
            <div
              class="text-custom-black dark:text-custom-white text-sm font-regular"
              style="margin-bottom: 4px;"
            >
              You can find the below information in the Details section on the OpenSea page of your NFT
            </div>
            <div
              class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10 mt-4"
            >
              <input
                autocomplete="off"
                class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                name="contractAddress"
                placeholder="Contract address"
                value=""
              />
            </div>
            <div
              class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10 mt-4"
            >
              <input
                autocomplete="off"
                class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                name="tokenId"
                placeholder="ID"
                value=""
              />
            </div>
            <div
              class="flex gap-2 absolute bottom-0 w-auto right-4 left-4"
            >
              <button
                class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full h-[44px] border-[1px] border-solid text-primary dark:text-custom-white border-primary dark:border-custom-white40 dark:hover:border-custom-white hover:border-primary-alt hover:text-primary-alt"
                type="button"
              >
                Actions.cancel
              </button>
              <button
                class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full disabled:text-custom-grey40 dark:disabled:text-custom-grey40 h-[44px] text-custom-white dark:text-custom-white bg-custom-grey10"
                disabled=""
                type="submit"
              >
                Import
              </button>
            </div>
          </div>
        </form>
        <div
          style="position: absolute; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
